libs/react-spa/shared/src/lib/validate.ts

```diff
@@ -0,0 +1,113 @@
+import { z } from 'zod'
+
+/**
+ * Creates one level deep object from a URLSearchParams
+ * Each key is split by dot notation and the value is the value of the key
+ *
+ * @example
+ * createSearchParamObj(new URLSearchParams('age.from=1&age.to=2&flightLeg.from=KEF&flightLeg.to=AER&period.from=2021-01-01&period.to=2021-01-02&state=active&state=inactive&isExplicit=true&airline=W6'))
+ * {
+ *   'age.from': 1,
+ *   'age.to': 2,
+ *   'flightLeg.from': 'KEF',
+ *   'flightLeg.to': 'AER',
+ *   'period.from': '2021-01-01',
+ *   'period.to': '2021-01-02',
+ *   state: ['active', 'inactive'],
+ *   isExplicit: 'true',
+ *   airline: 'W6',
+ * }
+ */
+export const createSearchParamsObj = (searchParams: URLSearchParams) => {
+  const obj: Record<string, unknown> = {}
+
+  for (const [objPath, value] of searchParams.entries()) {
+    if (!value) continue
+    const objValue = obj[objPath]
+
+    if (!objValue) {
+      obj[objPath] = value
+    } else if (objValue) {
+      obj[objPath] = Array.isArray(objValue)
+        ? [...objValue, value]
+        : [objValue, value]
+    }
+  }
+
+  return obj
+}
+
+/**
+ * Creates a deeply nested object from a one level deep object with dot notation
+ *
+ * @example
+ * const obj = {
+ *   'age.from': 1,
+ *   'age.to': 2,
+ *   'flightLeg.from': 'KEF',
+ *   'flightLeg.to': 'AER',
+ *   'period.from': '2021-01-01',
+ *   'period.to': '2021-01-02',
+ *   state: ['active', 'inactive'],
+ *   isExplicit: 'true',
+ *   airline: 'W6',
+ * }
+ * const result = dotNotationToNestedObject(obj)
+ * {
+ *   age: {
+ *     from: 1,
+ *     to: 2,
+ *   },
+ *   flightLeg: {
+ *     from: 'KEF',
+ *     to: 'AER',
+ *   },
+ *   period: {
+ *     from: '2021-01-01',
+ *     to: '2021-01-02',
+ *   },
+ *   state: ['active', 'inactive'],
+ *   isExplicit: true,
+ * }
+ */
+export const dotNotationToNestedObject = (obj: Record<string, unknown>) => {
+  return Object.entries(obj).reduce((parentAcc, [key, value]) => {
+    key
+      .split('.')
+      .reduce(
+        (acc, innerKey, index, arr) =>
+          acc[innerKey] || (acc[innerKey] = arr[index + 1] ? {} : value),
+        parentAcc as Record<string, any>,
+      )
+
+    return parentAcc
+  }, {} as Record<string, unknown>)
+}
+
+export const validateSearchParams = <T extends z.ZodTypeAny>({
+  request,
+  schema,
+}: {
+  request: Request
+  schema: T
+}) => {
+  const searchParams = new URL(request.url).searchParams
+  const values = createSearchParamsObj(searchParams)
+  const nestedObject = dotNotationToNestedObject(values)
+
+  return schema.parse(nestedObject) as z.infer<typeof schema>
+}
+
+export const validateFormData = async <T extends z.ZodTypeAny>({
+  request,
+  schema,
+}: {
+  request: Request
+  schema: T
+}) => {
+  const formData = await request.formData()
+  const values = Object.fromEntries(formData)
```

Comment:
Does this lose data if you pass the same property multiple times (state=a&state=b), expecting an array?

Reply:
Good spot. Fixed